Multiauth update user email address

maddiedierker opened the pull request:
For context, see spec [here](https://docs.google.com/document/d/1xg6pIrLMFFCkR_2v8w88OeZamr90hJXcdwnHv3xwNAw/edit?usp=sharing).

This PR refactors the `PATCH /users/email` endpoint to allow for migrated users to add an email address as their primary authentication option. Front-end changes for this will be in a separate PR.

islemaster commented:
:eyes: Looking now. Thanks for isolating the API changes!

islemaster commented on dashboard/app/models/user.rb:
```diff
@@ -830,6 +831,26 @@ def update_with_password(params, *options)
 
+  def update_primary_authentication_option(email)
+    # If an auth option exists with same email, set it to the user's primary authentication option
+    existing_auth_option = authentication_options.find {|ao| ao.email == email}
```
This logic doesn't handle students.  We'll also need to be able to accept only a hashed_email and find existing authentication options by matching hashed_email.

maddiedierker replied:
thanks for catching that! i'll update logic and add a test case to cover this

islemaster commented on dashboard/app/models/user.rb:
```diff
@@ -830,6 +831,26 @@ def update_with_password(params, *options)
+    existing_auth_option = authentication_options.find {|ao| ao.email == email}
+    if existing_auth_option
+      self.primary_authentication_option = existing_auth_option
+      return save
```
Scenario:
> Given I am a teacher with an email option and an oauth option, and the email option is primary,
> When I update my email address to match the oauth option
> Then my primary `AuthenticationOption` becomes the oauth option
> And (?) should my email `AuthenticationOption` be deleted?

maddiedierker replied:
i was thinking no, the existing email `AuthenticationOption` should stay as an available login option, but since the user is explicitly "updating" their email address, then that might be unintended behavior from the user's POV. 

so, instead the logic on lines 842-844 should always happen

islemaster replied:
Yeah, I'd do that for now.  Until we can make all of a user's login options visible in their account UI, I'd feel bad about keeping it around.

islemaster commented on dashboard/app/models/user.rb:
```diff
@@ -830,6 +831,26 @@ def update_with_password(params, *options)
+      email: email,
+      credential_type: 'email',
+      user: self
+    )
```
Is the `AuthenticationOption` persisted by the `User` `save` call?

maddiedierker replied:
yep, when the user saves it saves relational data too

islemaster commented on dashboard/app/views/devise/registrations/edit.html.haml:
```diff
@@ -170,4 +170,3 @@
     }.to_json
   }
 %script{src: minifiable_asset_path('js/devise/registrations/edit.js'), data: script_data}
-
```
Intentional?

maddiedierker replied:
nope, just my editor. i'll remove this

maddiedierker commented:
@islemaster i think i handled all of the comments you brought up, but let me know if you have any additional feedback!

islemaster commented on dashboard/app/controllers/registrations_controller.rb:
```diff
@@ -120,13 +120,27 @@ def set_email
+            false
+          end
+        else
+          current_user.update_primary_authentication_option(params[:user][:email])
```
This still needs to change.

When the user is under 13, `params[:user][:email]` will _never_ be populated, but they can still "change their email" by submitting a new hashed email to our server.

Here's the code on the client that omits email if the user is under 13.

https://github.com/code-dot-org/code-dot-org/blob/a1adf33ec94e7cb73a8be9b9eea3388f8888308a/apps/src/lib/ui/accounts/ChangeEmailController.js#L118-L119

I think our practice so far has been for the server to accept email, hashed_email, or both.
- If email is provided, the server generates a hashed email and saves both for a teacher, or just the hash for a student.
- If hashed email is provided, this is an error for a teacher, but can be used to update a student.
- If both are provided I think we trust the cleartext and the server regenerates the hash? Not sure though.

We can make some new choices about what we allow here, but we need to support the scenario where the client only sends a hash.

While we're at it, should we be using strong parameters here, or do we not need it since this is a fairly narrow case?

maddiedierker replied:
oh, i see. i thought that the email would always be submitted as plaintext. i'll update the endpoint and `update_primary_authentication_option` to account for a hashed email from the front-end.

we don't need strong params for this since it's already implemented:
https://github.com/code-dot-org/code-dot-org/blob/17e6a693350ce6ff841e834b574fee4408c15299/dashboard/app/controllers/registrations_controller.rb#L258-L263

islemaster replied:
We're not using that implementation though; notice that `set_email_params` is called in the non-migrated case below (new lines 139, 142) but not in the new migrated case, above.

maddiedierker replied:
oh, sorry about that. i was thinking of the `authenticate_scope!` before action, which i believe authenticates using strong params.

since the old endpoint was ultimately using the [devise method](https://github.com/code-dot-org/code-dot-org/blob/17e6a693350ce6ff841e834b574fee4408c15299/dashboard/app/controllers/registrations_controller.rb#L258-L263) `update_with_password` (or `update_without_password`, depending on the case), it made more sense to me to use strong params, whereas the new `update_primary_authentication_option` method can't do that since it requires more manual work and can't use the devise method.

maddiedierker replied:
although, since we define `set_email_params` and most likely won't be using `update_primary_authentication_method` elsewhere, maybe it does make more sense to still use strong params

maddiedierker commented:
@islemaster updated to account for receiving `hashed_email` from front-end. a few decisions i made from what you brought up:
- if both email and hashed_email are provided for a student, `update_primary_authentication_method` honors the plaintext email
- if a teacher tries to update their email address without the email param _or_ with the hashed_email param, `update_primary_authentication_method` will fail

maddiedierker commented:
thank you for taking the time to help! also, i just removed the constraint for teachers that `hashed_email` must be nil in user params--once i went to update the front-end, i realized that `hashed_email` was being supplied even for oauth-only teachers, so i removed the constraint. since the `hashed_email` param is being ignored anyways, there isn't a huge benefit in erroring if it does end up being supplied

ewjordan commented on dashboard/app/models/user.rb:
```diff
@@ -830,6 +831,31 @@ def update_with_password(params, *options)
+    params = {credential_type: 'email', user: self}
+    params[:email] = email unless email.nil?
+    params[:hashed_email] = hashed_email if email.nil?
+    self.primary_authentication_option = AuthenticationOption.new(params)
```
Will this actually save the new `AuthenticationOption` to the database? I'm not sure if saving the `User` will cascade to saving the `AuthenticationOption`

maddiedierker replied:
yep! it [autosaves](http://guides.rubyonrails.org/association_basics.html#options-for-belongs-to-autosave) dependencies. although i can't find if/where we're explicitly setting autosave, so i'll have to verify this in other environments besides dev because it could be different

ewjordan commented on dashboard/app/models/user.rb:
```diff
@@ -830,6 +831,31 @@ def update_with_password(params, *options)
     end
   end
 
+  def update_primary_authentication_option(user: {email: nil, hashed_email: nil})
```
When we add UI for options other than email, this will need to be updated, we'll probably want to think about how to best do this. I'm guessing at that point, we'll probably want to rename this so that it's an email-specific method, but for now this seems reasonable.

ewjordan commented on dashboard/test/controllers/registrations_controller_test.rb:
```diff
@@ -329,4 +329,150 @@ class RegistrationsControllerTest < ActionController::TestCase
+    teacher.reload
+    assert_response :success
+    assert_equal 'new@email.com', teacher.email
+  end
```
Great coverage in all of these tests, thank you!

ewjordan commented on dashboard/test/factories/factories.rb:
```diff
@@ -150,6 +150,58 @@
           end
         end
       end
+      trait :with_google_authentication_option do
```
Is it necessary to duplicate these traits inside `teacher` when they're already defined in `user`?

maddiedierker replied:
i defined this separately on the `teacher` and `student` factories, but you're right that it should just be on `user`. i'll make a cleanup PR with this

maddiedierker replied:
oops, nevermind, i just lost track of nesting in the factory! i'll remove the extraneous factory from `teacher`

ewjordan commented on dashboard/test/factories/factories.rb:
```diff
@@ -150,6 +150,58 @@
+            user: user,
+            email: user.email,
+            hashed_email: user.hashed_email,
+            credential_type: 'google_oauth',
```
`google_oauth` should actually be `google_oauth2`, which is a mistake I made when I wrote this code, and fix up in https://github.com/code-dot-org/code-dot-org/pull/23136/ - if we do need to duplicate these traits, it probably makes sense to fix this wherever they're duplicated.

maddiedierker replied:
i'll update this in a cleanup PR too!

ewjordan commented on dashboard/test/factories/factories.rb:
```diff
@@ -225,6 +277,22 @@
       end
     end
 
+    trait :with_migrated_google_authentication_option do
```
In https://github.com/code-dot-org/code-dot-org/pull/23136/ I have a `multi_auth_migrated` trait that serves a similar purpose, but I think the way you've done it here is better (in particular, the trait in my PR has to be applied in the right order, otherwise things fail), so I'll change my PR to match this approach.
